Template Updates 27 Jun 2023

mariolamassaavedra opened the pull request:
Updated event Templates for Origin alpha Trials

mariolamassaavedra committed:
Template Updates 27 Jun 2023

uakyol commented:
I added myself based on offline sync with Kate. Please let me know if you need more time working on it before a review

uakyol commented on src/main/proto/eventtemplates/video_event_template.proto:
```diff
@@ -30,15 +30,17 @@ message VideoEventTemplate {
-    uint32 completed_100_percent = 5;
+    option (.wfa.measurement.api.v2aplha.event_field) = {
+      display_name = "DigitalVideoCompletionStatus"
+    };
```
apparently there is a lint error here. Can you please have a look at the checks and fix them?

mariolamassaavedra replied:
Checked and ammended

uakyol commented:
The PR doesn't build. Can you please have a look at the "Build and test / Build and test (pull_request) " check and fix the errors.

uakyol reviewed (commented):
The PR doesn't build. Can you please have a look at the "Build and test / Build and test (pull_request) " check and fix the errors.

zasadnyy commented:
Hi @mariolamassaavedra , thanks for the PR.

For the Alpha pilots, I'd suggest to align event template protos with media types that we currently have in the UK Measurement Policy draft.

Namely I'd suggest the following structure:

```
common_template
- No fields

display_event_template
- Viewability

video_event_template
- Viewability
- DigitalVideoCompletionStatus

other_media_event_template
- No fields
```

This would allow to define explicitly in the event group schema what media types following event group would report.

Let me know what do you think.

khisba commented:
One last update from my side. We need to slightly amend the VIEWABLE_0_PERCENT_PLUS option.

      // More than 0% of the ad was displayed in the browser for more
      // than 0 seconds.
      0_PERCENT_PLUS = 1;

uakyol commented on src/main/proto/eventtemplates/common_template.proto:
```diff
@@ -32,18 +32,21 @@ message CommonTemplate {
+    enum Value {
+      // More than 0% of the ad was displayed in the browser for at least 1
+      // second for display ads or at least 2 seconds for video ads.
+      VIEWABLE_0_PERCENT_PLUS = 1;
```
Since this is now an enum, I think these should be 

VIEWABLE_0_PERCENT_TO_50_PERCENT = 1;
VIEWABLE_50_PERCENT_TO_100_PERCENT = 2;
VIEWABLE_100_PERCENT = 3;

Since only one of the values will be set.

khisba replied:
@uakyol - your suggestion doesn't match our use case. The enumeration options for viewability are intentionally overlapping. Viewability is a measurement standard which encompasses two elements:
1. % of the ad visible on-screen to the end user
2. amount of time that the ad was on-screen
 
When we select common_template.viewability == "0_PER_CENT_PLUS" in the filter then the expectation is that all impressions where greater than 0 pixels of the ad were viewable for greater than zero seconds should be returned. (i.e. not just impressions where between 0 and 50% of the ad was visible).


VIEWABLE_100_PER_CENT -- Return impressions where 100% of the ad was visible for at least 1 second for display ads or 2 seconds for video ads. AKA the new MRC viewability standard.

VIEWABLE_50_PER_CENT_PLUS -- Return impressions where greater than 50% of the ad was visible for at least 1 second for display ads and 2 seconds for video ads. AKA the old MRC viewabilty standard. 

0_PER_CENT_PLUS -- Return impressions where greater than 0% of the ad was visible for greater than 0 seconds. Note I've removed the prefix "VIEWABLE" from this option to indicate that this option is not an MRC Viewable definition and has a different time element.

It would also be possible to set no filter which should return all impressions regardless of viewability; and it would be possible to return "MRC Non-Viewable" events i.e. those that don't meet the MRC standard using a != operand. i.e. common_template.viewability != "VIEWABLE_100_PER_CENT".

uakyol replied:
@khisba - Lets talk more about the 2 dimensions and what should they exactly mean in the Thursday's call. But the issue with enums being _PLUS or _BETWEEN is not actually contradictory. 

These event templates are a schema for an event. So for this enum to have overlapping fields means for one event more than one of these enums have to be set at the same time. Which is not possible since they are enums.

If we change the field names to "ranges", we will achieve the same effect as you outlined above with these filters: 

**Previously VIEWABLE_100_PER_CENT** : 
eventFilter => common_template.viewability.value == VIEWABLE_100_PERCENT

**Previously VIEWABLE_50_PER_CENT_PLUS** : 
eventFilter => common_template.viewability.value == VIEWABLE_100_PERCENT || common_template.viewability.value == VIEWABLE_50_PERCENT_TO_100_PERCENT


**Previously 0_PER_CENT_PLUS** : 
eventFilter => common_template.viewability.value == VIEWABLE_100_PERCENT || common_template.viewability.value == VIEWABLE_50_PERCENT_TO_100_PERCENT || VIEWABLE_0_PERCENT_TO_50_PERCENT


**No eventFilter set** => Returns everything

khisba commented:
> Hi @mariolamassaavedra , thanks for the PR.
> 
> For the Alpha pilots, I'd suggest to align event template protos with media types that we currently have in the UK Measurement Policy draft.
> 
> Namely I'd suggest the following structure:
> 
> ```
> common_template
> - No fields
> 
> display_event_template
> - Viewability
> 
> video_event_template
> - Viewability
> - DigitalVideoCompletionStatus
> 
> other_media_event_template
> - No fields
> ```
> 
> This would allow to define explicitly in the event group schema what media types following event group would report.
> 
> Let me know what do you think.

@zasadnyy - This may be worth revisiting when we upgrade to the V2 Reporting Server. But, the reason for having a common template was to enable us to set a common filter on a ReportingSet which contains multiple EventGroups with different media types (i.e. common_template.viewability == "VIEWABLE_100_PER_CENT"). In the V1 Reporting Server, without this common template we would be forced to set individual filters per EventGroup (i.e. video_template.viewability == "VIEWABLE_100_PER_CENT" for video EventGroups, and display_template.viewability == "VIEWABLE_100_PER_CENT" for display EventGroups) which would then need to be set in the EventGroupUniverse to handle this case.

mariolamassaavedra reviewed (commented):
Fixed lint issue

mariolamassaavedra committed:
Fixed Lint issue

mariolamassaavedra committed:
Updated viewability option to be 0 plus

mariolamassaavedra committed:
Updated enum to start with consonant

mariolamassaavedra committed:
ADDED ENUM VLAUE FOR 0 TO COMPLY WITH PROTO3 ENUM SYNTAX

mariolamassaavedra committed:
Fixed Typo

mariolamassaavedra committed:
-

zasadnyy commented:
@khisba thanks for the context on the intentions for this design. I agree that suggested approach would make sense if we have only Display and Video media types but we have more.

Please consider that events falling into "other media" category do not have definition of the viewability. Therefore, viewability in the common template won't make sense for this media type.

Another benefit that I see from moving viewability into Display and Video templates - we'll be able to have separate enums for each type. This would allow us to clearly name viewability enums for Display and Video, eg `VIEWABLE_50_PER_CENT_PLUS_1_SECOND` and `VIEWABLE_50_PER_CENT_PLUS_2_SECONDS`

SanjayVas commented:
Reproducing this comment from the email thread:

> If we're planning on changing the UK pilot event templates now, I recommend the following fixes as well:
> 1. Drop the redundant `Template` suffix from the message name and the `_template` suffix from the template name.
>    * This also means renaming the fields in the `Event` message, since these must match the template names.
> 1. Drop the deprecated `event_field` annotation and use the new `template_field` annotation instead.
>    * This implies reducing the layers of nesting. There should only be top-level fields in event templates using the new annotation.
> 1. Use appropriate types rather than `uint32` for everything (e.g. `bool` or an `enum` type).
> 
> As a reminder, the `uk-pilot-event-templates` repo should be archived after the pilot.

I'd also recommend changing the protobuf package to something that would be globally unique. Maybe `halo.uk.pilot` with a corresponding Java package of `org.halo_cmm.uk.pilot`?

mariolamassaavedra committed:
Updated template to avoid deprecated definitions
Updated names to remove redundant suffixes
Updated Enums to Bool
Deleted event.proto as it doesn't apply to Origin

mariolamassaavedra committed:
Update bazel build file

mariolamassaavedra committed:
Options update

mariolamassaavedra committed:
Field Options

mariolamassaavedra committed:
Field Options 2

mariolamassaavedra committed:
C-lang formatting

mariolamassaavedra committed:
X

mariolamassaavedra committed:
C-lang formating

mariolamassaavedra committed:
bazel build updates

mariolamassaavedra committed:
Added other.proto
Removed viewability from common into both display and video

mariolamassaavedra committed:
C-lang formatting

mariolamassaavedra committed:
C-lang formatting

uakyol commented:
Looks good. Only thing is as @khisba mentioned, lets keep the file names consistent. Either with the _event suffix or without. 

I vote to keep out the _event suffix

mariolamassaavedra committed:
Updated build file format

mariolamassaavedra committed:
Build file update

zasadnyy commented on src/main/proto/eventtemplates/display_event.proto:
```diff
@@ -14,37 +14,32 @@
+    bool zero_percent_plus = 1;
+    // More than 0% of the ad was displayed in the browser for at least
+    // 1 second.
+    // AKA Old MRC Viewability standard
```
I believe for Display it's an actual MRC standard, for Video it was updated to be 100%. Lets remove this line

zasadnyy commented on src/main/proto/eventtemplates/display_event.proto:
```diff
@@ -14,37 +14,32 @@
+    // 1 second.
+    // AKA Old MRC Viewability standard
+    bool viewable_50_percent_plus = 2;
+    // More than 0% of the ad was displayed in the browser for at least
```
nit: "More than 100%"

zasadnyy commented on src/main/proto/eventtemplates/display_event.proto:
```diff
@@ -14,37 +14,32 @@
+    bool viewable_50_percent_plus = 2;
+    // More than 0% of the ad was displayed in the browser for at least
+    // 1 second.
+    // AKA New MRC Viewability standard
```
ditto: For Display standard didn't change, let's remove this line

zasadnyy commented on src/main/proto/eventtemplates/other.proto:
```diff
@@ -14,31 +14,17 @@
 option java_multiple_files = true;
 
-message VideoEventTemplate {
+message Other {
```
+1 to Ugur's comment, I'd suggest to align naming, eg call this "OtherMediaEvent"

Same applies to the file name

zasadnyy commented on src/main/proto/eventtemplates/video_event.proto:
```diff
@@ -0,0 +1,55 @@
+    // 0 seconds.
+    bool zero_percent_plus = 1;
+    // More than 50% of the ad was displayed in the browser for at least
+    // 2 seconds.
```
nit: more precise would be _"... for at least 2 continuous seconds"_

zasadnyy commented on src/main/proto/eventtemplates/video_event.proto:
```diff
@@ -0,0 +1,55 @@
+    // AKA Old MRC Viewability standard
+    bool viewable_50_percent_plus = 2;
+    // 100% of the ad was displayed in the browser for at least
+    // 2 seconds.
```
Same here, _"... for at least 2 continuous seconds"_

zasadnyy reviewed (commented):
Did one more pass, overall looks good, just added a few small suggestions

mariolamassaavedra committed:
Renamed files to follow _event format
Updated build file to compile common.proto as a dependency
Added common.proto to all templates for future implementation
Updated MRC definitions

zasadnyy reviewed (approved):
Looks good, thanks

uakyol commented:
Sorry, I think there is a misunderstanding here. I hadn't realized you got rid of the wrapper `Event` message. We have to have that to disambiguate for fields in different templates. For example we have viewability in both DisplayEvent and VideoEvent. The edps wouldn't know which one to filter against. [Here is the documentation](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/aafaca5fe6c6d5827af22ba09e8651365f0d89c5/src/main/proto/wfa/measurement/api/v2alpha/event_annotations.proto#L41) : `Name of the field in an event message that uses this template. Required. Immutable.`

Here is I think the best way to define that wrapper message : 

```
message Event {
  Display display = 1;
  Video video = 2;
  Other other = 3;
}
```

Note that I got rid of `_event` suffixes since they are not events but templates.

mariolamassaavedra committed:
Updates after alignments with Ugur

mariolamassaavedra committed:
Typo fixed

mariolamassaavedra committed:
Removed extra commas

mariolamassaavedra committed:
Updated file names

mariolamassaavedra committed:
Naming update

mariolamassaavedra committed:
Updates to naming conventions

mariolamassaavedra committed:
c-land formatting for bazel build file

mariolamassaavedra committed:
c-lang formatting added

mariolamassaavedra committed:
c-lanf reformat

mariolamassaavedra committed:
Fixed Typo non ascii character

mariolamassaavedra committed:
Typo

mariolamassaavedra committed:
XX

zasadnyy reviewed (approved):
Looks good, thanks

SanjayVas commented on src/main/proto/eventtemplates/display_template.proto:
```diff
@@ -0,0 +1,44 @@
+
+message Display {
+  option (.wfa.measurement.api.v2alpha.event_template) = {
+    name: "display_template"
```
This name **must** match the field name in the event message.

mariolamassaavedra replied:
Thanks for the review Sanjay. I thought it was the message name (line 24) which was being refered. I've change all the _template.proto files template names so that we remove the _template as indeed becomes redundant

mariolamassaavedra committed:
Removed _template from the template option name value
